File: app/src/main/java/org/wikipedia/dataclient/mwapi/MwResponse.java
package org.wikipedia.dataclient.mwapi;

import com.google.gson.annotations.SerializedName;

import org.wikipedia.json.PostProcessingTypeAdapter;
import org.wikipedia.model.BaseModel;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public abstract class MwResponse extends BaseModel implements PostProcessingTypeAdapter.PostProcessable {
    @SuppressWarnings({"unused"}) @Nullable private List<MwServiceError> errors;
    @SuppressWarnings("unused,NullableProblems") @SerializedName("servedby") @NonNull private String servedBy;

    @Override
    public void postProcess() {
        if (errors != null && !errors.isEmpty()) {
            throw new MwException(errors.get(0));
        }
    }
}

File: app/src/main/java/org/wikipedia/dataclient/mwapi/MwResponse.kt
package org.wikipedia.dataclient.mwapi

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
abstract class MwResponse {
    private val errors: List<MwServiceError>? = null

    @SerialName("servedby")
    private val servedBy: String? = null

    init {
        if (errors?.isNotEmpty() == true) {
            // prioritize "blocked" or "abusefilter" errors over others.
            errors.firstOrNull { it.key.contains("blocked") || it.key.startsWith("abusefilter-") }?.let {
                throw MwException(it)
            }
            errors.firstOrNull { it.key.contains("assertuserfailed") || it.key.contains("notloggedin") }?.let {
                throw MwNotLoggedInException(it)
            }
            throw MwException(errors.first())
        }
    }
}
